connmgr/conngating.go

```diff
@@ -0,0 +1,64 @@
+package connmgr
+
+import (
+	"github.com/libp2p/go-libp2p-core/control"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/transport"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// ConnectionGater can be implemented by a type that supports active
+// inbound or outbound connection gating.
+//
+// A ConnectionGater will be consulted during different states in the life-cycle of a connection and
+// the specific gating function that will be called depends on the lifecycle state of the
+// connection. Hence, it is important to implement this interface keeping in mind the specific life cycle state
+// at which you'd like to gate/block the connection.
+//
+// `InterceptDial` and `InterceptPeerDial` are called when we try an outbound dial.
+//
+// `InterceptAccept` is called after we've accepted an inbound connection from a socket but before we
+// begin upgrading it.
+//
+// `InterceptSecured` is called for inbound and outbound connections after we've negotiated
+// the security protocol to use for the connection.
+//
+// `InterceptUpgraded` is called for inbound and outbound connections after we've finished upgrading
+// a connection to have both security and stream multiplexing.
+//
+// This feature can be used to implement *strict* connection management
+// behaviours, such as hard limiting of connections once a max count has been
+// reached.
+type ConnectionGater interface {
+	// InterceptDial tests whether we're permitted to dial the specified multiaddr.
+	// Insofar filter.Filters is concerned, this would map to its AddrBlock method,
+	// with the inverse condition.
+	// This is to be called by the network/swarm when dialling.
+	InterceptDial(ma.Multiaddr) (allow bool)
+
+	// InterceptPeerDial tests whether we're permitted to Dial the specified peer.
+	// This is to be called by the network/swarm when dialling.
+	InterceptPeerDial(p peer.ID) (allow bool)
```

Comment:
@raulk I've added this function to gate outbound dials based on the peerId. Otherwise, users would have to wait till the dial is complete and InterceptSecured is called. 

Also, gives this functionality to transports that do not use an upgrader.

Reply:
Not really necessary. On an outbound dial, the multiaddr contains the peer ID.